From 67489aa03be86b063a40d9c5d6cc77f05cacd36a Mon Sep 17 00:00:00 2001
From: "Ni, Erchang" <erchang.ni@intel.com>
Date: Tue, 30 Apr 2024 04:20:54 +0000
Subject: [PATCH 51/54] put SBL parameters to the head of commandline

The old code append abl_cmd_line to the end of cmd_conf. Now we prepand abl_cmd_line to cmd_conf.

Tracked-On: OAM-117889
Signed-off-by: Ni, Erchang <erchang.ni@intel.com>
Signed-off-by: Chen Gang G <gang.g.chen@intel.com>
---
 libkernelflinger/android.c | 27 ++++++++++++++++++++++++---
 1 file changed, 24 insertions(+), 3 deletions(-)

diff --git a/libkernelflinger/android.c b/libkernelflinger/android.c
index 0c97415..f1673bc 100644
--- a/libkernelflinger/android.c
+++ b/libkernelflinger/android.c
@@ -1561,14 +1561,35 @@ static EFI_STATUS setup_command_line(
                 cmd_conf[cmdlen] = 0;
         }
 
-        /* append command line from ABL */
+        /* prepand command line from ABL */
         if (abl_cmd_len > 0)
         {
-                cmd_conf[cmdlen] = ' ';
-                ret = memcpy_s(cmd_conf + cmdlen + 1, abl_cmd_len + 1, abl_cmd_line, abl_cmd_len + 1);
+                CHAR8* new_cmd_conf = AllocatePool(cmdsize);
+                if (new_cmd_conf == NULL) {
+                        ret = EFI_OUT_OF_RESOURCES;
+                        goto out;
+                }
+
+                ret = memcpy_s(new_cmd_conf, abl_cmd_len, abl_cmd_line, abl_cmd_len);
                 if (EFI_ERROR(ret)) {
+                        FreePool(new_cmd_conf);
                         goto out;
                 }
+
+                if (cmdlen > 0) {
+                        new_cmd_conf[abl_cmd_len] = ' ';
+                        ret = memcpy_s(new_cmd_conf + abl_cmd_len + 1, cmdlen, cmd_conf, cmdlen);
+                        if (EFI_ERROR(ret)) {
+                                FreePool(new_cmd_conf);
+                                goto out;
+                        }
+                }
+
+                if (cmd_conf != NULL) {
+                        FreePool(cmd_conf);
+                }
+
+                cmd_conf = new_cmd_conf;
                 cmdlen += abl_cmd_len + 1;
                 cmd_conf[cmdlen] = '\0';
         }
-- 
2.25.1

